File: mobile/lib/features/reviews/domain/entities/review.dart
import 'package:flutter/foundation.dart';

/// The moderation state of a guest review.
///
/// There is **no backend review domain in the MVP** — this vocabulary is the
/// contract the mobile app needs and that a future guest review endpoint must
/// provide (see `mobile/docs/mobile-phase-10-loyalty-reviews.md`). The app
/// never assumes a review is publicly visible: only [published] means "live".
enum ReviewStatus {
  /// Submitted, awaiting moderation. NOT publicly visible.
  pending('pending'),

  /// Approved by moderation and live.
  published('published'),

  /// Rejected by moderation. The guest may be allowed to submit again — the
  /// backend decides.
  rejected('rejected');

  const ReviewStatus(this.wireValue);

  final String wireValue;

  static ReviewStatus fromWire(String? value) {
    for (final ReviewStatus s in ReviewStatus.values) {
      if (s.wireValue == value) return s;
    }
    return ReviewStatus.pending;
  }

  bool get isPending => this == ReviewStatus.pending;
  bool get isPublished => this == ReviewStatus.published;
  bool get isRejected => this == ReviewStatus.rejected;
}

/// A guest review for one completed/stayed reservation.
///
/// One review per eligible reservation. Numeric rating 1–5, optional text.
/// No sub-ratings (room / cleanliness / service / location / staff) — none are
/// in any approved contract.
@immutable
class Review {
  const Review({
    required this.id,
    required this.reservationId,
    required this.rating,
    required this.status,
    this.text,
    this.createdAt,
  });

  final String id;
  final String reservationId;

  /// 1–5, integer.
  final int rating;

  /// Optional free text. Never whitespace-only (trimmed at the boundary).
  final String? text;

  final ReviewStatus status;
  final DateTime? createdAt;

  bool get hasText => text != null && text!.trim().isNotEmpty;

  @override
  bool operator ==(Object other) =>
      other is Review &&
      other.id == id &&
      other.reservationId == reservationId &&
      other.rating == rating &&
      other.text == text &&
      other.status == status &&
      other.createdAt == createdAt;

  @override
  int get hashCode =>
      Object.hash(id, reservationId, rating, text, status, createdAt);

  @override
  String toString() => 'Review($id, r$rating, ${status.wireValue})';
}

File: backend/app/Domain/Review/Models/Review.php
<?php

namespace App\Domain\Review\Models;

use App\Domain\HotelGroup\Models\Hotel;
use App\Domain\IdentityAccess\Models\User;
use App\Domain\Reservation\Models\Guest;
use App\Domain\Reservation\Models\Reservation;
use Database\Factories\ReviewFactory;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * A guest's post-stay review of one completed reservation
 * (mobile/docs/mobile-phase-10-loyalty-reviews.md). One per reservation —
 * enforced by the `reviews.reservation_id` UNIQUE constraint, not just the
 * service layer.
 *
 * Only `published` reviews are ever shown to another guest; `pending` /
 * `rejected` are visible to their own author and to staff only.
 */
class Review extends Model
{
    use HasFactory;

    public const STATUS_PENDING = 'pending';

    public const STATUS_PUBLISHED = 'published';

    public const STATUS_REJECTED = 'rejected';

    protected $fillable = [
        'reservation_id',
        'guest_id',
        'hotel_id',
        'rating',
        'text',
        'status',
        'moderated_by_user_id',
        'moderated_at',
    ];

    protected function casts(): array
    {
        return [
            'rating' => 'integer',
            'moderated_at' => 'datetime',
        ];
    }

    public function reservation(): BelongsTo
    {
        return $this->belongsTo(Reservation::class);
    }

    public function guest(): BelongsTo
    {
        return $this->belongsTo(Guest::class);
    }

    public function hotel(): BelongsTo
    {
        return $this->belongsTo(Hotel::class);
    }

    public function moderatedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'moderated_by_user_id');
    }

    protected static function newFactory(): ReviewFactory
    {
        return ReviewFactory::new();
    }
}
